server/src/main/java/org/apache/cassandra/sidecar/acl/authorization/PermissionFactoryImpl.java

```
@@ -0,0 +1,100 @@
+
+import static org.apache.cassandra.sidecar.acl.authorization.DomainAwarePermission.WILDCARD_PART_DIVIDER_TOKEN;
+import static org.apache.cassandra.sidecar.acl.authorization.FeaturePermission.ALL_FEATURE_PERMISSIONS;
+
+/**
+ * Implementation of {@link PermissionFactory}
+ */
+public class PermissionFactoryImpl implements PermissionFactory
+{
+    private static final String FORBIDDEN_FEATURE_PERMISSION_PREFIX = "*:";
+    private final List<CompositePermission> supportedFeaturePermissions;
+
+    public PermissionFactoryImpl()
+    {
+        this(Collections.unmodifiableList(ALL_FEATURE_PERMISSIONS));
+    }
+
+    /**
+     * Creates a {@link PermissionFactoryImpl} given a list of supported feature permissions.
+     *
+     * @param supportedFeaturePermissions list of feature permissions sidecar recognizes and honors
+     */
+    public PermissionFactoryImpl(List<CompositePermission> supportedFeaturePermissions)
+    {
+        this.supportedFeaturePermissions = supportedFeaturePermissions;
+    }
+
+    /**
+     * @return list of supported feature permissions.
+     */
+    public List<CompositePermission> supportedFeaturePermissions()
+    {
+        return supportedFeaturePermissions;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Permission createPermission(String name)
+    {
+        Objects.requireNonNull(name, "name cannot be null");
+        Permission permission = createFeaturePermission(name);
+        if (permission != null)
+        {
+            return permission;
+        }
+        boolean isDomainAware = name.contains(WILDCARD_PART_DIVIDER_TOKEN);
+        return isDomainAware ? new DomainAwarePermission(name) : new StandardPermission(name);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompositePermission createFeaturePermission(String name)
+    {
+        if (name.startsWith(FORBIDDEN_FEATURE_PERMISSION_PREFIX))
+        {
+            return null;
```

Throw `IllegalArgumentException` instead of returning null. `*:` is not a permission that can be permitted by `DomainAwarePermission` neither. Let's fail fast.

---

@yifan-c I returned null, because we try to create a feature permission if not fallback to creating a basic permission. Not throwing an exception, allows someone to override `DomainAwarePermission` to allow for `*:` prefix in their permission, without having to modify the `PermissionFactoryImpl`. When we have fallback in place, didn't want to throw prematurely.

---

Can you show me, in this patch, how to return the domain aware permission subclass w/o having a new factory implementation? 
If someone wants to modifies the behavior in domain aware permission validation, they also need to have a new factory implementation. TBH, the imaginary scenario is out of the scope.

---

Moved the exception throwing to `createPermission` itself, that way it is not a feature permission specific exception.

---

they would need to override just `createPermission` they can still keep `createFeaturePermission` as is.